src/validation.cpp

```diff
@@ -3222,6 +3253,16 @@ static bool FindBlockPos(FlatFilePos &pos, unsigned int nAddSize, unsigned int n
 
     if (!fKnown) {
         while (vinfoBlockFile[nFile].nSize + nAddSize >= MAX_BLOCKFILE_SIZE) {
+            // when the undo file is keeping up with the block file, we flush it explicitly, here
+            // when it is lagging behind (more blocks arrive than are being connected), we let the
+            // undo block write case handle it
+            LogPrintf("iterating nFile: last height = %d; chain tip height = %d -> %s\n",
+                vinfoBlockFile[nFile].nHeightLast,
+                ChainActive().Tip()->nHeight,
+                vinfoBlockFile[nFile].nHeightLast == ChainActive().Tip()->nHeight ? "FLUSH UNDO FILE" : "DO NOT FLUSH UNDO FILE");
+            if (vinfoBlockFile[nFile].nHeightLast == ChainActive().Tip()->nHeight) {
+                FlushUndoFile(nFile, true);
```

Comment:
This will flush the undo file before the corresponding block file which is flushed a [few lines below](https://github.com/bitcoin/bitcoin/pull/17994/files#diff-24efdb00bfbe56b140fb006b562cc70bR3279). Also visible in the log you pasted where `finalize undo 2005` appears before `Leaving block file 2005`.

I am not sure but I think we want to always flush the block file first and the undo file afterwards.

Reply:
I *think* the way it works now is fine; the "unflushed undo files" output is consistently showing the current opened one only, so it indicates the same.

Let me go through the steps:

* Sync state: HEIGHT = 100, BLK = 100, UNDO = 100, block file 0
* Block 101 comes in (we assume it does not fit into the current block file, so we will move to block file 1)
* First we want to store it to disk: 

https://github.com/bitcoin/bitcoin/blob/c4400e1b8a62269d04b57a46d68941fdbaae4303/src/validation.cpp#L3731-L3736

* We go find a position and end up in the next block file (1), at which point we flush the undo file (0):

https://github.com/bitcoin/bitcoin/blob/c4400e1b8a62269d04b57a46d68941fdbaae4303/src/validation.cpp#L3254-L3264

* Sync state: HEIGHT = 100, BLK = 101, UNDO = 100, block file 1
* We now want to accept block 101, and at this point we write the undo data (for block 101)

https://github.com/bitcoin/bitcoin/blob/c4400e1b8a62269d04b57a46d68941fdbaae4303/src/validation.cpp#L1773-L1787

* The condition does not trigger here, because we are synced up (as you pointed out), but we end up writing into undo file 1, and undo file 0 was finalized as it should.

Let me know if I got something wrong in the above (or if there are any other holes in the patch). I will update the log above once it moves to block file 2007.

Reply:
Yes, the above is correct (or at least coincides with my understanding for this elusive code!) - I think the code will do what you describe and that it is the correct behavior wrt finalizing the undo.

However, one piece is missing from the above - when do we flush block file 0?
> We go find a position and end up in the next block file (1), at which point we flush the undo file (0)

... and _after this_ we flush the blk file 0.

Just to be explicit, something like this may be warranted, delay the flush of the undo until after we have flushed the block file:

```diff
+    int thisUndoNeedsFlushingAndFinalizing = -1;
+
     if (!fKnown) {
         while (vinfoBlockFile[nFile].nSize + nAddSize >= MAX_BLOCKFILE_SIZE) {
             // when the undo file is keeping up with the block file, we flush it explicitly, here
             // when it is lagging behind (more blocks arrive than are being connected), we let the
             // undo block write case handle it
             LogPrintf("iterating nFile: last height = %d; chain tip height = %d -> %s\n",
                 vinfoBlockFile[nFile].nHeightLast,
                 ChainActive().Tip()->nHeight,
                 vinfoBlockFile[nFile].nHeightLast == ChainActive().Tip()->nHeight ? "FLUSH UNDO FILE" : "DO NOT FLUSH UNDO FILE");
             if (vinfoBlockFile[nFile].nHeightLast == ChainActive().Tip()->nHeight) {
-                FlushUndoFile(nFile, true);
+                thisUndoNeedsFlushingAndFinalizing = nFile;
             }
             nFile++;
             if (vinfoBlockFile.size() <= nFile) {
                 vinfoBlockFile.resize(nFile + 1);
             }
         }
         pos.nFile = nFile;
         pos.nPos = vinfoBlockFile[nFile].nSize;
     }
 
     if ((int)nFile != nLastBlockFile) {
         if (!fKnown) {
             LogPrintf("Leaving block file %i: %s\n", nLastBlockFile, vinfoBlockFile[nLastBlockFile].ToString());
         }
+        // Order is important: first FlushBlockFile(), then FlushUndoFile().
+        // FlushBlockFile() will flush internally nLastBlockFile and at this
+        // point I believe thisUndoNeedsFlushingAndFinalizing == nLastBlockFile
+        // so we will flush corresponding blk and undo files.
         FlushBlockFile(!fKnown);
+        if (thisUndoNeedsFlushingAndFinalizing != -1) {
+            FlushUndoFile(thisUndoNeedsFlushingAndFinalizing, true);
+        }
         nLastBlockFile = nFile;
     }
```

Reply:
Oh wait, I see what you mean -- you're concerned about the non-finalized undo flush happening after the finalized undo flush (called from within the finalized block flush), right?

I don't think this it's a problem to flush twice. The only difference between a finalized=true flush and a finalized=false flush is L87 in

https://github.com/bitcoin/bitcoin/blob/a3abe25f720104e4abf0d3f01f0b847929167b8b/src/flatfile.cpp#L81-L98

which means the non-finalized flush is a NOP, I think.

Reply:
No, sorry, I mentioned my concern (order of flushing of blk and undo files), but did not mention why is that a concern. Previously the code always flushed blkN before flushing revN (finalizing is not relevant for this):

https://github.com/bitcoin/bitcoin/blob/97b0687501cee77a9170f9e288755a5d268e9bd4/src/validation.cpp#L1747-L1748

In case of a crash (or power failure) between the two, we would end up with a block data on disk without a corresponding undo data. I presume this is ok in the current code.

Now, with the change in this PR we would flush revN first and blkN afterwards. This could potentially leave us with an undo data for some block(s) without the corresponding block data. There is at least this code which would get upset by that:

https://github.com/bitcoin/bitcoin/blob/97b0687501cee77a9170f9e288755a5d268e9bd4/src/validation.cpp#L4810

Now, I am not 100% sure if that is a real concern or not, but is relatively easy to tweak in the current patch so that it keeps flushing first blk and then undo (as before the patch).